File: chaeyun/2_tree-solutions/is-balanced-tree.py

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root) -> bool:
        def height(root):
            if root == None:
                return 0
            
            h_left = height(root.left)
            if h_left < 0:
                return -1
            h_right = height(root.right)
            if h_right < 0 or abs(h_left - h_right) > 1:
                return -1

            return max(h_left, h_right) + 1
        
        return height(root) >= 0

File: chaeyun/2_tree-solutions/is-balanced-tree.cpp
#include <algorithm>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    int height(TreeNode* root) {
        if (!root) {
            return 0;
        }
        int hLeft = height(root->left);
        if (hLeft < 0) {
            return -1;
        }
        int hRight = height(root->right);
        if (hRight < 0 || std::abs(hLeft - hRight) > 1) {
            return -1;
        }
        return std::max(hLeft, hRight) + 1;
    }

    bool isBalanced(TreeNode* root) {
        return height(root) >= 0;
    }
};